### DISCUSSION TITLE
Error in gap heat transfer based mortar

### DISCUSSION THREAD

**Discussion by jin0123456** [idaholab/moose#27683] [Q&A Modules: Thermal Hydraulics]
Hello,

I want to use the existing function provided in MOOSE to solve a gap heat transfer problem. As my problem involves contact heat transfer so it needs coupling `Solid Mechanics` module to acquire contact pressure. Directly I chose motar way, but it gave me the error
`The requested mortar interface AutomaticMortarGeneration object does not yet exist!`
I put my input and mesh there. https://github.com/jin0123456/didactic-pancake
I wonder if it's my mesh problem. In motar way is there a requirement that the lengths of the two sides match? As my sides don't match.
Much appreciated for advice.

**Comment by GiudGiud** [comment]
@gambka

**Comment by gambka** [comment]
Your setup looks to have the thermal contact model `test_thm.i` as the Master app with the mechanical contact calculation (`test_mech.i`) as the subapp. Here, you are passing the `interface_normal_lm` through a transfer from the sub to master app. I've not seen this type of setup before for coupled mechanical and thermal contact. Is there any reason you do not have both mechanical and thermal contact in the same input file as is done in the test `moose/modules/combined/test/tests/gap_heat_transfer_mortar/finite-2d/varied_pressure_thermomechanical_mortar.i`? Have you verified that your simulation runs when it all in one input? If so, then I suspect it is related the mesh creation or the multiapp transfers. I cannot run your inputs as there are several non-MOOSE objects. If you have a MOOSE only example that generates the same error using your multiapp setup, it would be easier to debug.

For the meshing, you create the lower D mesh blocks manually in the master app, but in the sub app they are created automatically by the mechanical contact `[Contact]` block. Are these blocks consistent? It may be that the master app is requesting the `interface_normal_lm` which has not yet been created in the sub app yet. To answer your question about differing lengths of the sidesets in contact, it is fine to have sides of different lengths. I recommend making the primary surface the longer surface in your 2D-RZ setup.

**Reply by jin0123456** [reply]
Thanks for youe advice about the primary surface setting. For your questions,
1.I don't have both mechanical and thermal contact in the same input file as it's not converged in the case. I have to split it into two apps.
2.I have tried this in a test before and it went right. I put them there as closed_gap_thm.i and closed_gap_mech.i.
https://github.com/jin0123456/didactic-pancake
3.I suppose the problem is not `interface_normal_lm` transfer as I didn't apply `side_contact` in `thermal_contact`.
The related settings about contact are the same in my case and test so I suppose it's the mesh problem?

**Reply by lindsayad** [reply]
The coupling between thermal and mechanical contact is often quite tight. You are unlikely to get better overall simulation convergence by splitting them between multiapps. If you are having issues converging thermomechnical contact in a single input, then I think we need to focus our debugging there, not on the multiapp case

**Reply by jin0123456** [reply]
OK, I will try it in a single input.
